// azure-event-hubs/dist/lib/amqp-common/log.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const debugModule = require("debug");
/**
 * @ignore
 * log statements for cbs
 */
exports.cbs = debugModule("azure:amqp-common:cbs");
/**
 * @ignore
 * log statements for error
 */
exports.error = debugModule("azure:amqp-common:error");
/**
 * @ignore
 * log statements for datatransformer
 */
exports.transformer = debugModule("azure:amqp-common:datatransformer");
/**
 * @ignore
 * log statements for messageHeader
 */
exports.msgHeader = debugModule("azure:amqp-common:messageHeader");
/**
 * @ignore
 * log statements for messageProperties
 */
exports.msgProperties = debugModule("azure:amqp-common:messageProperties");
/**
 * @ignore
 * log statements for retry
 */
exports.retry = debugModule("azure:amqp-common:retry");
/**
 * @ignore
 * log statements for reqreslink
 */
exports.reqres = debugModule("azure:amqp-common:reqreslink");
