// azure-event-hubs/dist/lib/amqp-common/messageHeader.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const log = require("./log");
/**
 * Describes the operations that can be performed on the message header.
 * @module MessageHeader
 */
var MessageHeader;
(function (MessageHeader) {
    /**
     * Converts MessageHeader to AmqpMessageHeader.
     *
     * @param {MessageHeader} props Message header.
     * @returns {AmqpMessageHeader} AmqpMessageHeader
     */
    function toAmqpMessageHeader(props) {
        const amqpHeader = {
            delivery_count: props.deliveryCount,
            durable: props.durable,
            first_acquirer: props.firstAcquirer,
            priority: props.priority,
            ttl: props.ttl
        };
        log.msgHeader("To AmqpMessageHeader: %O", amqpHeader);
        return amqpHeader;
    }
    MessageHeader.toAmqpMessageHeader = toAmqpMessageHeader;
    /**
     * Converts AmqpMessageHeader to MessageHeader.
     *
     * @param {AmqpMessageHeader} props Amqp Message Header
     * @returns {MessageHeader} MessageHeader.
     */
    function fromAmqpMessageProperties(props) {
        const msgHeader = {
            deliveryCount: props.delivery_count,
            durable: props.durable,
            firstAcquirer: props.first_acquirer,
            priority: props.priority,
            ttl: props.ttl
        };
        log.msgHeader("From AmqpMessageHeader: %O", msgHeader);
        return msgHeader;
    }
    MessageHeader.fromAmqpMessageProperties = fromAmqpMessageProperties;
})(MessageHeader = exports.MessageHeader || (exports.MessageHeader = {}));
