// azure-event-hubs/dist/lib/connectionContext.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const os = require("os");
const process = require("process");
const uuid = require("uuid/v4");
const log = require("./log");
const constants_1 = require("./util/constants");
const amqp_common_1 = require("./amqp-common");
const managementClient_1 = require("./managementClient");
const rhea_promise_1 = require("./rhea-promise");
const constants_2 = require("./amqp-common/util/constants");
var ConnectionContext;
(function (ConnectionContext) {
    const userAgent = "/js-event-hubs";
    function getUserAgent(options) {
        const finalUserAgent = options.userAgent ? `${userAgent},${options.userAgent}` : userAgent;
        if (finalUserAgent.length > constants_2.maxUserAgentLength) {
            throw new Error(`The user-agent string cannot be more than 128 characters in length.` +
                `The given user-agent string is: ${finalUserAgent} with length: ${finalUserAgent.length}`);
        }
        return finalUserAgent;
    }
    ConnectionContext.getUserAgent = getUserAgent;
    function create(config, options) {
        if (!options)
            options = {};
        amqp_common_1.ConnectionConfig.validate(config, { isEntityPathRequired: true });
        const packageVersion = constants_1.packageJsonInfo.version;
        const connectionOptions = {
            transport: amqp_common_1.Constants.TLS,
            host: config.host,
            hostname: config.host,
            username: config.sharedAccessKeyName,
            port: 5671,
            reconnect: false,
            properties: {
                product: "MSJSClient",
                version: packageVersion,
                platform: `(${os.arch()}-${os.type()}-${os.release()})`,
                framework: `Node/${process.version}`,
                "user-agent": getUserAgent(options)
            }
        };
        const connection = new rhea_promise_1.Connection(connectionOptions);
        const connectionLock = `${amqp_common_1.Constants.establishConnection}-${uuid()}`;
        const connectionContext = {
            wasConnectionCloseCalled: false,
            connectionLock: connectionLock,
            negotiateClaimLock: `${amqp_common_1.Constants.negotiateClaim}-${uuid()}`,
            connection: connection,
            connectionId: connection.id,
            cbsSession: new amqp_common_1.CbsClient(connection, connectionLock),
            config: config,
            tokenProvider: options.tokenProvider ||
                new amqp_common_1.SasTokenProvider(config.endpoint, config.sharedAccessKeyName, config.sharedAccessKey),
            senders: {},
            receivers: {},
            dataTransformer: options.dataTransformer || new amqp_common_1.DefaultDataTransformer()
        };
        const mOptions = {
            address: options.managementSessionAddress,
            audience: options.managementSessionAudience
        };
        connectionContext.managementSession = new managementClient_1.ManagementClient(connectionContext, mOptions);
        // register handlers on the connection.
        const onConnectionOpen = (context) => {
            connectionContext.wasConnectionCloseCalled = false;
            log.context("[%s] setting 'wasConnectionCloseCalled' property of connection context to %s.", connectionContext.connection.id, connectionContext.wasConnectionCloseCalled);
        };
        connectionContext.connection.registerHandler(rhea_promise_1.ConnectionEvents.connectionOpen, onConnectionOpen);
        const disconnected = async (context) => {
            const connectionError = context.connection && context.connection.error
                ? context.connection.error
                : undefined;
            if (connectionError) {
                log.error("[%s] Error (context.connection.error) occurred on the amqp connection: %O", connectionContext.connection.id, connectionError);
            }
            const contextError = context.error;
            if (contextError) {
                log.error("[%s] Error (context.error) occurred on the amqp connection: %O", connectionContext.connection.id, contextError);
            }
            const state = {
                wasConnectionCloseCalled: connectionContext.wasConnectionCloseCalled,
                numSenders: Object.keys(connectionContext.senders).length,
                numReceivers: Object.keys(connectionContext.receivers).length
            };
            // The connection should always be brought back up if the sdk did not call connection.close()
            // and there was atleast one sender/receiver link on the connection before it went down.
            log.error("[%s] state: %O", connectionContext.connection.id, state);
            if (!state.wasConnectionCloseCalled && (state.numSenders || state.numReceivers)) {
                log.error("[%s] connection.close() was not called from the sdk and there were some " +
                    "sender or receiver links or both. We should reconnect.", connectionContext.connection.id);
                await amqp_common_1.delay(constants_2.connectionReconnectDelay);
                // reconnect senders if any
                for (const sender of Object.values(connectionContext.senders)) {
                    if (!sender.isConnecting) {
                        log.error("[%s] calling detached on sender '%s' with address '%s'.", connectionContext.connection.id, sender.name, sender.address);
                        sender.detached().catch((err) => {
                            log.error("[%s] An error occurred while reconnecting the sender '%s' with adress '%s' %O.", connectionContext.connection.id, sender.name, sender.address, err);
                        });
                    }
                    else {
                        log.error("[%s] sender '%s' with address '%s' is already reconnecting. Hence not " +
                            "calling detached on the sender.", connectionContext.connection.id, sender.name, sender.address);
                    }
                }
                // reconnect receivers if any
                for (const receiver of Object.values(connectionContext.receivers)) {
                    if (!receiver.isConnecting) {
                        log.error("[%s] calling detached on receiver '%s' with address '%s'.", connectionContext.connection.id, receiver.name, receiver.address);
                        receiver.detached().catch((err) => {
                            log.error("[%s] An error occurred while reconnecting the receiver '%s' with adress '%s' %O.", connectionContext.connection.id, receiver.name, receiver.address, err);
                        });
                    }
                    else {
                        log.error("[%s] receiver '%s' with address '%s' is already reconnecting. Hence not " +
                            "calling detached on the receiver.", connectionContext.connection.id, receiver.name, receiver.address);
                    }
                }
            }
        };
        connectionContext.connection.registerHandler(rhea_promise_1.ConnectionEvents.disconnected, disconnected);
        log.context("[%s] Created connection context successfully.", connectionContext.connectionId);
        return connectionContext;
    }
    ConnectionContext.create = create;
})(ConnectionContext = exports.ConnectionContext || (exports.ConnectionContext = {}));
