// azure-event-hubs/dist/lib/log.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const debugModule = require("debug");
/**
 * @ignore
 * log statements for error
 */
exports.error = debugModule("azure:event-hubs:error");
/**
 * @ignore
 * log statements for management
 */
exports.mgmt = debugModule("azure:event-hubs:management");
/**
 * @ignore
 * log statements for sender
 */
exports.sender = debugModule("azure:event-hubs:sender");
/**
 * @ignore
 * log statements for receiver
 */
exports.receiver = debugModule("azure:event-hubs:receiver");
/**
 * @ignore
 * log statements for receiverbatching
 */
exports.batching = debugModule("azure:event-hubs:receiverbatching");
/**
 * @ignore
 * log statements for receiverstreaming
 */
exports.streaming = debugModule("azure:event-hubs:receiverstreaming");
/**
 * @ignore
 * log statements for linkEntity
 */
exports.link = debugModule("azure:event-hubs:linkEntity");
/**
 * @ignore
 * log statements for connectionContext
 */
exports.context = debugModule("azure:event-hubs:connectionContext");
/**
 * @ignore
 * log statements for client
 */
exports.client = debugModule("azure:event-hubs:client");
/**
 * @ignore
 * log statements for iothub client
 */
exports.iotClient = debugModule("azure:event-hubs:iothubClient");
